### ISSUE TITLE
[Bug/Support]: failing to resolve sufficiently high version of dependencies for eglot (eldoc, json-rpc)

### ISSUE THREAD

**Issue by petergaultney** [progfolio/elpaca#341]
### Confirmation

- [X] I have checked the documentation (README, Wiki, docstrings, etc)
- [ ] I am  checking these without reading them.
- [X] I have searched previous issues to see if my question is a duplicate.

### Elpaca Version

Elpaca edc0dc2 grafted, HEAD -> master, origin/master, origin/HEAD
installer:      0.7
emacs-version:  GNU Emacs 29.4 (build 1, aarch64-apple-darwin23.4.0)
 of 2024-06-22
git --version:  git version 2.41.0

### Operating System

MacOS 14.6.1

### Description

I'm trying elpaca for the first time. I found [this StackOverflow issue](https://emacs.stackexchange.com/questions/81555/elpaca-problems-with-eglot-version-and-flymake), which helped me get past the first instance of this problem, which was a `failed` state for `eglot` with the message that it requires `eldoc` 1.14, and i had 1.13. 

Anyway, adding 
```
(use-package eldoc
  :config
  (provide 'upgraded-eldoc))
;; elpaca seems unable to install eglot without eldoc 1.14... and won't resolve that on its own for whatever reason.
```

to my `python.el` right before the `(use-package eglot ...` seemed to get past this error (i watched elpaca install a newer version of eldoc). 

But then it happened again with `json-rpc` - the exact same error.

<img width="1408" alt="Screenshot 2024-08-21 at 11 33 40 AM" src="https://github.com/user-attachments/assets/7cc130a8-dc97-4778-a026-5ceb8f4e4461">

I'm happy to have a workaround - but I guess I also don't really understand the underlying issue here - if elpaca can detect that I need a new version... why doesn't it install that version for me?

I am very interested in getting this to work... `straight.el` was becoming very hard for me to understand - I am hoping this will be a better experience overall.

**Comment by progfolio** [comment]
Thanks for taking the time to fill out a support ticket. 

> I found [this StackOverflow issue](https://emacs.stackexchange.com/questions/81555/elpaca-problems-with-eglot-version-and-flymake), which helped me get past the first instance of this problem, which was a `failed` state for `eglot` with the message that it requires `eldoc` 1.14, and i had 1.13. 
> 
> Anyway, adding 
> ```
> (use-package eldoc
>   :config
>   (provide 'upgraded-eldoc))

The answer is inaccurate. There is no need to provide an "upgraded-eldoc" feature.

Either of the following should suffice:

```emacs-lisp
;; without use-package
(elpaca eldoc)
;;or with elpaca-use-package integration
(use-package eldoc :ensure t)
```
In general, I don't recommend stack overflow as a support channel for Elpaca. I dont have an account there and don't plan on farming enough "reputation" points to be able to answer questions there. This issue tracker is the authority. 

> to my `python.el` right before the `(use-package eglot ...` seemed to get past this error (i watched elpaca install a newer version of eldoc). 

I don't know what "python.el" is.
Im assuming that's a file you're loading in your init.el. Correct? 

> But then it happened again with `json-rpc` - the exact same error.

If you'd like to install the latest version of json-rpc, the same advice given above applies.

> I'm happy to have a workaround - but I guess I also don't really understand the underlying issue here - if elpaca can detect that I need a new version... why doesn't it install that version for me?

Because installing a newer version of a package (especially a "core" Emacs package) can break other packages. I can't make the decision for users whether that's what they want. It must be decided on a case by case basis.

My general advice is to update Emacs itself if you find yourself needing to update a bunch of core packages. Another solution is to use an older version of the package which requires all the updated dependencies. Does that help?

**Comment by petergaultney** [comment]
all of this helps a lot, yes. 

I think my core confusion was not knowing (😳 ) that `eldoc` and `jsonrpc` were built-in emacs packages. I think I ended up figuring that out... maybe by reading other issues in this repo? I'm not actually sure how I should have discovered that, but I think I mostly know how to figure that out in the future.

I was getting totally different errors when I tried `(use-package eldoc :ensure t)` and `(use-package jsonrpc :ensure t)` - i will make a separate issue for those once I'm available later.
